File: lib/providers/github.ts
/**
 * GitHub public-source discovery.
 *
 * GitHub's REST search API is free and reachable without auth at
 * 60 req/hr per IP. With a personal-access token in `GITHUB_TOKEN` we
 * get 5000 req/hr — strongly recommended for production, but the path
 * works either way.
 *
 * We hit /search/users to find people matching the query, then look up
 * each user's profile for bio/company/location so the chat agent can
 * preview real candidates instead of just usernames.
 */

import type { ProspectCandidate } from "./brave-search"

const SEARCH_URL = "https://api.github.com/search/users"
const USER_URL = "https://api.github.com/users"

interface GhSearchHit {
  login: string
  html_url: string
}

interface GhUser {
  login: string
  name: string | null
  bio: string | null
  company: string | null
  location: string | null
  blog: string | null
  html_url: string
}

function authHeaders(): HeadersInit {
  const base: HeadersInit = {
    Accept: "application/vnd.github+json",
    "User-Agent": "leadgenai/0.5",
  }
  if (process.env.GITHUB_TOKEN) {
    return { ...base, Authorization: `Bearer ${process.env.GITHUB_TOKEN}` }
  }
  return base
}

export async function searchGithubUsers(
  query: string,
  maxResults: number,
): Promise<ProspectCandidate[]> {
  // GitHub's search-user query syntax: keyword + qualifiers
  // Example: "founder location:bangalore" finds users whose profile
  // mentions "founder" and lives in Bangalore.
  const url = `${SEARCH_URL}?q=${encodeURIComponent(query)}&per_page=${Math.min(maxResults, 30)}`
  let res: Response
  try {
    res = await fetch(url, {
      headers: authHeaders(),
      signal: AbortSignal.timeout(10_000),
    })
  } catch {
    return []
  }
  if (!res.ok) return []
  const data = (await res.json()) as { items?: GhSearchHit[] }
  const hits = data.items?.slice(0, maxResults) ?? []

  // Fetch full profiles in parallel (3 at a time — be polite).
  const profiles = await mapConcurrent(hits, 3, async (h) => {
    try {
      const r = await fetch(`${USER_URL}/${h.login}`, {
        headers: authHeaders(),
        signal: AbortSignal.timeout(8_000),
      })
      if (!r.ok) return null
      return (await r.json()) as GhUser
    } catch {
      return null
    }
  })

  const out: ProspectCandidate[] = []
  for (const p of profiles) {
    if (!p) continue
    out.push({
      name: p.name ?? p.login,
      title: p.bio ?? "GitHub user",
      company: (p.company ?? "").replace(/^@/, "") || "(independent)",
      location: p.location ?? undefined,
      source: "github",
      source_url: p.html_url,
      snippet: [p.bio, p.location, p.company]
        .filter(Boolean)
        .join(" · ") || `GitHub profile @${p.login}`,
    })
  }
  return out
}

// Tiny concurrency-limited map — duplicated here so the provider has
// no internal cross-import on the agent layer.
async function mapConcurrent<T, R>(
  items: T[],
  concurrency: number,
  fn: (item: T) => Promise<R>,
): Promise<R[]> {
  const out: R[] = new Array(items.length)
  let next = 0
  const workers: Promise<void>[] = []
  for (let i = 0; i < Math.min(concurrency, items.length); i++) {
    workers.push(
      (async () => {
        while (true) {
          const idx = next++
          if (idx >= items.length) return
          out[idx] = await fn(items[idx])
        }
      })(),
    )
  }
  await Promise.all(workers)
  return out
}

File: backend-python/leadgen_backend/providers/github.py
import os
import httpx
import asyncio

SEARCH_URL = "https://api.github.com/search/users"
USER_URL = "https://api.github.com/users"

def auth_headers() -> dict:
    base = {
        "Accept": "application/vnd.github+json",
        "User-Agent": "leadgenai/0.5"
    }
    token = os.getenv("GITHUB_TOKEN")
    if token:
        base["Authorization"] = f"Bearer {token}"
    return base

async def search_github_users(query: str, max_results: int) -> list[dict]:
    params = {"q": query, "per_page": min(max_results, 30)}
    try:
        async with httpx.AsyncClient(timeout=10) as client:
            res = await client.get(SEARCH_URL, headers=auth_headers(), params=params)
            if res.status_code >= 400: return []
            data = res.json()
    except httpx.HTTPError:
        return []
        
    hits = data.get("items", [])[:max_results]
    
    # fetch profiles concurrently
    async def fetch_profile(client, login: str):
        try:
            r = await client.get(f"{USER_URL}/{login}", headers=auth_headers())
            if r.status_code < 400:
                return r.json()
        except httpx.HTTPError:
            pass
        return None

    async with httpx.AsyncClient(timeout=8) as client:
        tasks = [fetch_profile(client, hit["login"]) for hit in hits]
        # In TS we used a bounded concurrency map, asyncio.gather is fine for max 30 profiles.
        profiles = await asyncio.gather(*tasks)
        
    out = []
    for p in profiles:
        if not p: continue
        name = p.get("name") or p.get("login", "")
        title = p.get("bio") or "GitHub user"
        comp = p.get("company", "") or ""
        comp = comp.lstrip("@") or "(independent)"
        loc = p.get("location")
        
        snip_parts = [p.get("bio"), loc, comp]
        snippet = " · ".join([str(x) for x in snip_parts if x]) or f"GitHub profile @{p.get('login', '')}"
        
        out.append({
            "name": name,
            "title": title,
            "company": comp,
            "location": loc,
            "source": "github",
            "source_url": p.get("html_url", ""),
            "snippet": snippet,
        })
        
    return out
